cmd/timewarp/README.md

```diff
@@ -0,0 +1,130 @@
+filtering responses to only include packages and versions that existed at the
+specified point in time. Currently supported registries:
+
+- NPM (https://registry.npmjs.org/)
+- PyPI (https://pypi.org/)
+
+## Installation
+
+```bash
+go install github.com/google/oss-rebuild/cmd/timewarp@latest
+```
+
+Or clone the repository and build:
+
+```bash
+git clone https://github.com/google/oss-rebuild.git
+cd oss-rebuild
+go build ./cmd/timewarp
+```
+
+## Usage
+
+Start the Timewarp server:
+
+```bash
+timewarp --port 8081
+```
+
+The server will listen on the specified port (default: 8081).
+
+### Using with NPM
+
+Set the NPM registry to point to Timewarp, including the desired timestamp in RFC3339 format:
+
+```bash
+npm --registry "http://npm:2024-09-13T10:31:26.370Z@localhost:8081" install express
+```
+
+Or set it as an environment variable:
+
+```bash
+export npm_config_registry="http://npm:2024-09-13T10:31:26.370Z@localhost:8081"
+npm install express
+```
+
+### Using with pip/PyPI
+
+Set the PyPI index URL to point to Timewarp, including the desired timestamp in RFC3339 format:
+
+```bash
+pip install --index-url "http://pypi:2013-12-23T07:45:10.417Z@localhost:8081/" requests
+```
+
+Or set it as an environment variable:
+
+```bash
+export PIP_INDEX_URL="http://pypi:2013-12-23T07:45:10.417Z@localhost:8081/"
+pip install requests
+```
+
+### Using with curl
+
+You can also use curl to directly query the timewarp service:
+
+```bash
+curl -u "npm:2024-09-13T10:31:26.370Z" http://localhost:8081/express | jq | less
+```
+
+```bash
+curl -u "pypi:2013-12-23T07:45:10.417Z" http://localhost:8081/pypi/requests/json | jq | less
+```
+
+## Design
+
+Timewarp uses the HTTP Basic Authentication mechanism to pass both the platform type and target timestamp. The username field specifies the registry type (`npm` or `pypi`), and the password field contains the RFC3339 timestamp for the desired point in time.
+
+When a request comes in, Timewarp:
+
+1. Parses the platform and timestamp from Basic Auth credentials
+2. Forwards the request to the appropriate upstream registry
+3. For JSON responses, filters out versions published after the specified time
+4. Returns the modified response to the client
+
+### Details
+
+- Timewarp maintains no state and doesn't cache responses
+- For NPM packages, it:
+  - Filters out versions published after the specified time
+  - Updates the latest version tag to point to the latest version before the cutoff
+  - Updates metadata like repository and description to match the new latest version
+- For PyPI packages, it:
+  - Filters out versions published after the specified time
+  - Removes individual files uploaded after the time cutoff
+  - Fetches and merges version-specific data for the new latest version
+
+### Limitations
+
+- Time format must be RFC3339 (e.g., `2023-05-13T10:31:26.370Z`)
+- The timestamp must be between January 1, 2000 and 24 hours in the future
```

Comment:
24 hours in the future of what? current time?

Reply:
eh actually I'm just not gonna document that part. i initially added that to catch mistakes but i'm not sure that still has value (and is just an odd behavior to communicate). i'll remove it in a follow-on